### PR TITLE
[JEP441] Add support for switch pattern matching

### LINKED ISSUE #4343: Pattern matching for switch - JEP 441
This is a fully supported feature in Java 21 (no longer a preview)

https://openjdk.org/jeps/441

### PR DESCRIPTION
**Pull request by johannescoetzee**
Switch pattern matching changes introduced in https://openjdk.org/jeps/441

This PR adds support for switch pattern matching with guards and resolves https://github.com/javaparser/javaparser/issues/4343 using the approach discussed in https://github.com/javaparser/javaparser/issues/4361 and partially implemented in https://github.com/javaparser/javaparser/pull/4357.

In particular, it adds support for the new `CasePattern [Guard]` syntax in 
```
SwitchLabel:
  case CaseConstant {, CaseConstant}
  case null [, default]
  case CasePattern [Guard]    // <- New support
  default

CasePattern:
  Pattern

Guard:
  when Expression
```

The main changes in this PR are:
* Syntax support in the parser
* Syntax support in the concrete syntax model
* Addition of an optional `SwitchEntry.guard` field which contains the guard expression if any
* Type solver checks switch pattern labels to find variable declarations
* Fixes for the PrettyPrinter and LexicalPreservingPrinter
* Some unit tests for all of the above.

### CODE REVIEW

File: javaparser-core-serialization/src/test/java/com/github/javaparser/serialization/JavaParserJsonSerializerTest.java [hunk 1]
```diff
@@ -43,7 +43,7 @@ void test() {
...
...e\":{\"!\":\"com.github.javaparser.ast.expr.SimpleName\",\"range\":{\"beginLine\":1,\"beginColumn\":14,\"endLine\":1,\"endColumn\":17},\"tokenRange\":{\"beginToken\":{\"kind\":97,\"text\":\"util\"},\"endToken\":{\"kind\":97,\"text\":\"util\"}},\"identifier\":\"util\"},\"scope\":{\"!\":\"com.github.javaparser.ast.type.ClassOrInterfaceType\",\"range\":{\"beginLine\":1,\"beginColumn\":9,\"endLine\":1,\"endColumn\":12},\"tokenRange\":{\"beginToken\":{\"kind\":97,\"text\":\"java\"},\"endToken\":{\"kind\":97,\"text\":\"java\"}},\"name\":{\"!\":\"com.github.javaparser.ast.expr.SimpleName\",\"range\":{\"beginLine\":1,\"beginColumn\":9,\"endLine\":1,\"endColumn\":12},\"tokenRange\":{\"beginToken\":{\"kind\":97,\"text\":\"java\"},\"endToken\":{\"kind\":97,\"text\":\"java\"}},\"identifier\":\"java\"},\"annotations\":[]},\"annotations\":[]},\"annotations\":[]}}],\"annotations\":[]}],\"modifiers\":[],\"name\":{\"!\":\"com.github.javaparser.ast.expr.SimpleName\",\"range\":{\"beginLine\":1,\"beginColumn\":7,\"endLine\":1,\"endColumn\":7},\"tokenRange\":{\"beginToken\":{\"kind\":97,\"text\":\"X\"},\"endToken\":{\"kind\":97,\"text\":\"X\"}},\"identifier\":\"X\"},\"annotations\":[]}]}", serialized);
```
Comment (johannescoetzee) [inline javaparser-core-serialization/src/test/java/com/github/javaparser/serialization/JavaParserJsonSerializerTest.java:46]: The changes here are just some token kind values which changed due to the addition of the `WHEN` token

File: javaparser-core/src/main/java/com/github/javaparser/JavaToken.java [hunk 2]
```diff
@@ -336,78 +334,79 @@ public enum Kind {
         EXPORTS(76),
         PROVIDES(77),
         TRANSITIVE(78),
-        LONG_LITERAL(79),
```
Comment (johannescoetzee) [inline javaparser-core/src/main/java/com/github/javaparser/JavaToken.java:339]: These changes are generated

File: javaparser-core/src/main/java/com/github/javaparser/JavaToken.java [hunk 3]
```diff
@@ -417,150 +416,152 @@ public enum Kind {
 
         public static Kind valueOf(int kind) {
             switch(kind) {
-                case 150:
```
Comment (johannescoetzee) [inline javaparser-core/src/main/java/com/github/javaparser/JavaToken.java:420]: These changes are generated

**Comment by johannescoetzee** [comment]
I just realized I forgot to update the validators. I'll update the PR with that probably this afternoon.

File: javaparser-core/src/main/java/com/github/javaparser/ast/stmt/SwitchEntry.java [hunk 4]
```diff
@@ -202,6 +217,12 @@ public boolean remove(Node node) {
         if (node == null) {
             return false;
         }
+        if (guard != null) {
```
Comment (jlerbsc) [inline javaparser-core/src/main/java/com/github/javaparser/ast/stmt/SwitchEntry.java:219]: Why this and the counterpart?
Do you have a test case that tests the deletion of a node with a guard expression?

File: javaparser-core/src/main/java/com/github/javaparser/ast/visitor/CloneVisitor.java [hunk 5]
```diff
@@ -258,6 +257,7 @@ public Visitable visit(final Parameter n, final Object arg) {
     @Override
     public Visitable visit(final InitializerDeclaration n, final Object arg) {
         BlockStmt body = cloneNode(n.getBody(), arg);
+        NodeList<AnnotationExpr> annotations = cloneList(n.getAnnotations(), arg);
```
Comment (jlerbsc) [inline javaparser-core/src/main/java/com/github/javaparser/ast/visitor/CloneVisitor.java:260]: How does this relate to the subject of the PR?

File: javaparser-core/src/main/java/com/github/javaparser/ast/visitor/GenericVisitorAdapter.java [hunk 6]
```diff
@@ -1418,6 +1418,11 @@ public R visit(final SuperExpr n, final A arg) {
     @Override
     public R visit(final SwitchEntry n, final A arg) {
         R result;
+        if (n.getGuard().isPresent()) {
```
Comment (jlerbsc) [inline javaparser-core/src/main/java/com/github/javaparser/ast/visitor/GenericVisitorAdapter.java:1421]: I don't understand this block!

File: javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/resolution/SwitchExprTest.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    private CompilationUnit parse(String code) {
+        TypeSolver typeSolver = new ReflectionTypeSolver();
+        ParserConfiguration parserConfiguration = new ParserConfiguration();
+        parserConfiguration.setSymbolResolver(new JavaSymbolSolver(typeSolver));
+        parserConfiguration.setLanguageLevel(ParserConfiguration.LanguageLevel.BLEEDING_EDGE);
+        return new JavaParser(parserConfiguration).parse(code).getResult().get();
+    }
+
+    @Test
+    public void todoremove() {
```
Comment (jlerbsc) [inline javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/resolution/SwitchExprTest.java]: Why this method name?

File: javaparser-core/src/main/java/com/github/javaparser/printer/ConcreteSyntaxModel.java [hunk 8]
```diff
@@ -160,7 +160,7 @@ private static CsmElement typeArguments() {
...
...s.put(SwitchEntry.class, sequence(comment(), conditional(SWITCH_STATEMENT_ENTRY, FLAG, conditional(ObservableProperty.LABELS, IS_NOT_EMPTY, sequence(token(GeneratedJavaParserConstants.CASE), space(), list(ObservableProperty.LABELS), token(GeneratedJavaParserConstants.COLON)), sequence(token(GeneratedJavaParserConstants._DEFAULT), token(GeneratedJavaParserConstants.COLON))), conditional(ObservableProperty.LABELS, IS_NOT_EMPTY, conditional(ObservableProperty.DEFAULT, FLAG, sequence(token(GeneratedJavaParserConstants.CASE), space(), list(ObservableProperty.LABELS), comma(), space(), token(GeneratedJavaParserConstants._DEFAULT), space(), token(GeneratedJavaParserConstants.ARROW)), sequence(token(GeneratedJavaParserConstants.CASE), space(), list(ObservableProperty.LABELS), conditional(ObservableProperty.GUARDED_ENTRY, FLAG, sequence(space(), token(GeneratedJavaParserConstants.WHEN), space(), child(ObservableProperty.GUARD))), space(), token(GeneratedJavaParserConstants.ARROW))), sequence(token(GeneratedJavaParserConstants._DEFAULT), space(), token(GeneratedJavaParserConstants.ARROW)))), newline(), indent(), list(ObservableProperty.STATEMENTS, newline(), none(), newline()), unindent()));
```
Comment (jlerbsc) [inline javaparser-core/src/main/java/com/github/javaparser/printer/ConcreteSyntaxModel.java]: I'm not sure that the GUARDED_ENTRY property is necessary. Can't we just use the GUARD property?

File: javaparser-core/src/main/java/com/github/javaparser/ast/stmt/SwitchEntry.java [hunk 9]
```diff
@@ -140,6 +146,15 @@ public boolean isSwitchStatementEntry() {
         return type == Type.STATEMENT_GROUP;
     }
 
+    /**
+     * This is required for the ConcreteSyntaxModel for deciding if a "when" token should
+     * be expected or not.
+     */
+    @DerivedProperty
+    public boolean isGuardedEntry() {
```
Comment (jlerbsc) [inline javaparser-core/src/main/java/com/github/javaparser/ast/stmt/SwitchEntry.java]: I'm not sure that the GUARDED_ENTRY property is necessary. Can't we just use the GUARD property?

**Comment by jlerbsc** [comment]
Great work. I've got a few questions, but it seems fine to me, with the possible exception of the derived property introduced for the Lexical Preserving Printer.

**Comment by johannescoetzee** [comment]
I've updated this PR with the following:
* Removed the `GUARDED_ENTRY` property in favour of using the `IS_PRESENT` condition on the `GUARD`
* Updated validators with tests
* Added removal tests

**Comment by jlerbsc** [comment]
A job well done. Thanks for your time.

**Comment by pativa** [comment]
This looks great and should resolve an issue we are facing with javaparser!

Do you have any estimate as of when this will be released? 🙏

**Comment by jlerbsc** [comment]
This will be integrated into the next release, which will provide support for all Java 21 non-preview features

**Pull request merged**